### PR TITLE
#104: Added broadcast collective operation

### LINKED ISSUE #104: Add broadcast collective operation
# Background

- The ML Frameworks need to exchange connection information between all UDF instances of a UDF call to build a cluster for distributed machine learning
- To send information from one UDF instance to all other UDF instances, we need to implement the [Broadcast](https://en.wikipedia.org/wiki/Collective_operation#Broadcast) pattern 
- Tips: 
  - Count the rounds of collective operations and add them to the messages in case of strugglers or resends.

### PR DESCRIPTION
**Pull request by tkilias**
Fixes #104

### CODE REVIEW

File: exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+from structlog.typing import FilteringBoundLogger
+
+from exasol_advanced_analytics_framework.udf_communication import messages
+from exasol_advanced_analytics_framework.udf_communication.messages import Gather
+from exasol_advanced_analytics_framework.udf_communication.peer import Peer
+from exasol_advanced_analytics_framework.udf_communication.peer_communicator import PeerCommunicator
+from exasol_advanced_analytics_framework.udf_communication.serialization import serialize_message, deserialize_message
+from exasol_advanced_analytics_framework.udf_communication.socket_factory.abstract import SocketFactory, Frame
+
+LOGGER: FilteringBoundLogger = structlog.getLogger()
```
Comment (Nicoretti) [inline exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py:13]:
```suggestion
_LOGGER: FilteringBoundLogger = structlog.getLogger()
```

File: exasol_advanced_analytics_framework/udf_communication/communicator.py [hunk 2]
```diff
@@ -95,6 +97,14 @@ def gather(self, value: bytes) -> Optional[List[bytes]]:
                                  number_of_instances_per_node=self._number_of_instances_per_node)
         return gather()
 
+    def broadcast(self, value: Optional[bytes]) -> bytes:
+        sequence_number = self._next_sequence_number()
+        broadcast = BroadcastOperation(sequence_number=sequence_number, value=value,
+                                       localhost_communicator=self._localhost_communicator,
+                                       multi_node_communicator=self._multi_node_communicator,
+                                       socket_factory=self._socket_factory)
+        return broadcast()
```
Comment (Nicoretti) [inline exasol_advanced_analytics_framework/udf_communication/communicator.py:106]:
```suggestion
    def broadcast(self, value: Optional[bytes]) -> bytes:
        sequence_number = self._next_sequence_number()
        operation = BroadcastOperation(sequence_number=sequence_number, value=value,
                                       localhost_communicator=self._localhost_communicator,
                                       multi_node_communicator=self._multi_node_communicator,
                                       socket_factory=self._socket_factory)
        return operation()
```

File: exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    def _send_messages_to_local_peers_from_multi_node_leaders(self):
+        self._logger.info("_send_messages_to_local_peers_from_multi_node_leaders")
+        leader = self._localhost_communicator.leader
+        for peer in self._localhost_communicator.peers():
+            if peer != leader:
+                value_frame = self._socket_factory.create_frame(self._value)
+                frames = self._construct_broadcast_message(destination=peer,
+                                                           leader=leader,
+                                                           value_frame=value_frame)
+                self._localhost_communicator.send(peer=peer, message=frames)
```
Comment (Nicoretti) [inline exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py:102]:
```suggestion
        for peer in [p for p in self._localhost_communicator.peers() if p != leader]:
            value_frame = self._socket_factory.create_frame(self._value)
            frames = self._construct_broadcast_message(
                destination=peer,
                leader=leader,
                value_frame=value_frame
            )
            self._localhost_communicator.send(peer=peer, message=frames)
```

File: exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        if self._multi_node_communicator is not None:
+            self._logger.info("_send_messages_to_local_leaders")
+            leader = self._multi_node_communicator.leader
+            for peer in self._multi_node_communicator.peers():
+                if peer != leader:
+                    value_frame = self._socket_factory.create_frame(self._value)
+                    frames = self._construct_broadcast_message(destination=peer,
+                                                               leader=leader,
+                                                               value_frame=value_frame)
+                    self._multi_node_communicator.send(peer=peer, message=frames)
```
Comment (Nicoretti) [inline exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py:91]:
```suggestion
    def _send_messages_to_local_leaders(self):
        if self._multi_node_communicator is None:
            return

        self._logger.info("_send_messages_to_local_leaders")
        leader = self._multi_node_communicator.leader
        peers = [peer for peer in self._multi_node_communicator.peers() if peer != leader]

        for peer in peers:
            value_frame = self._socket_factory.create_frame(self._value)
            frames = self._construct_broadcast_message(
                destination=peer,
                leader=leader,
                value_frame=value_frame
            )
            self._multi_node_communicator.send(peer=peer, message=frames)
```

File: exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        self._logger.info("_forward_from_multi_node_leader")
+        value_frame = self.receive_value_frame_from_multi_node_leader()
+        leader = self._localhost_communicator.leader
+        for peer in self._localhost_communicator.peers():
+            if peer != leader:
+                frames = self._construct_broadcast_message(destination=peer,
+                                                           leader=leader,
+                                                           value_frame=value_frame)
+                self._localhost_communicator.send(peer=peer, message=frames)
+        return value_frame.to_bytes()
```
Comment (Nicoretti) [inline exasol_advanced_analytics_framework/udf_communication/broadcast_operation.py:65]:
```suggestion
    def _forward_from_multi_node_leader(self) -> bytes:
        self._logger.info("_forward_from_multi_node_leader")
        value_frame = self.receive_value_frame_from_multi_node_leader()
        leader = self._localhost_communicator.leader
        peers = [peer in self._localhost_communicator.peers() if peer != leader]

        for peer in peers:
            frames = self._construct_broadcast_message(
                destination=peer,
                leader=leader,
                value_frame=value_frame
            )
            self._localhost_communicator.send(peer=peer, message=frames)

        return value_frame.to_bytes()
```

File: tests/integration_tests/without_db/udf_communication/test_broadcast.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    BidirectionalQueue, assert_processes_finish, \
+    CommunicatorTestProcessParameter
+
+structlog.configure(
+    context_class=dict,
+    logger_factory=WriteLoggerFactory(file=Path(__file__).with_suffix(".log").open("wt")),
+    processors=[
+        structlog.contextvars.merge_contextvars,
+        ConditionalMethodDropper(method_name="debug"),
+        # ConditionalMethodDropper(method_name="info"),
```
Comment (Nicoretti) [inline tests/integration_tests/without_db/udf_communication/test_broadcast.py:26]:
```suggestion
```

File: .github/workflows/integration_tests_with_db.yaml [hunk 7]
```diff
@@ -11,6 +11,18 @@ jobs:
     runs-on: ubuntu-latest
```
Comment (Nicoretti) [inline .github/workflows/integration_tests_with_db.yaml:11]: maybe we could use a slimmer image like `alpine`?

File: .github/workflows/integration_tests_with_db.yaml [hunk 8]
```diff
@@ -11,6 +11,18 @@ jobs:
...
+        uses: jlumbroso/free-disk-space@v1.3.0
+        with:
+          tool-cache: false
+          android: true
+          dotnet: true
+          haskell: true
+          large-packages: true
+          docker-images: true
+          swap-storage: false
+
```
Comment (Nicoretti) [inline .github/workflows/integration_tests_with_db.yaml:25]: Did you list/request this action also [here](https://teamexasol.slack.com/archives/G8SB06C2K/p1697461011560019)?

**Review by Nicoretti — APPROVED** [review]

File: .github/workflows/integration_tests_with_db.yaml (hunk 7 above)
Comment (tkilias) [reply .github/workflows/integration_tests_with_db.yaml:11]: Not a docker image, only GitHub provided images supported and they add all the garbage

Comment (Nicoretti) [reply .github/workflows/integration_tests_with_db.yaml:11]: right, my bad :grimacing:. sry cache was still occupied by the previous context :sweat_smile:

**Pull request merged**